### PR TITLE
refactor(es6): Upgrade AutomaticPrefetchPlugin to es6

### PR DESCRIPTION
**Pull request by chicoxyzzy**
**What kind of change does this PR introduce?**

Refactor

**Did you add tests for your changes?**

Existing Tests are passing

**If relevant, link to documentation update:**

N/A

**Summary**

update AutomaticPrefetchPlugin to ES6

**Does this PR introduce a breaking change?**

No

### CODE REVIEW

**Comment by chicoxyzzy** [comment]
Not sure why Appveyor failed ¯\_(ツ)_/¯

**Comment by chicoxyzzy** [comment]
Also not sure why it passes now O_o

**Comment by sokra** [comment]
I restarted the appveyor build. Seems like that one test case is flanky.

**Comment by TheLarkInn** [comment]
@chicoxyzzy Would you be willing to write tests for it in this PR or a merge-dependent separate PR? Otherwise, if not I'll create an issue and someone can pick it up if they want.

**Comment by chicoxyzzy** [comment]
@TheLarkInn from what I read in https://github.com/webpack/webpack/issues/1566 and https://github.com/webpack/webpack.js.org/issues/126 I don't understand how to test this plugin :(

**Comment by TheLarkInn** [comment]
@chicoxyzzy Okay so I can help with that I believe. So on `Compilation.js` there is a `.prefetch` function which adds a dependency to the moduleChain. 

Typically it goes in dependency order but I believe let's you jump the order on certain deps. that may take longer. 

This used to be done by using `PrefetchPlugin()` and manually specifying the dependencies. It looks like in this logics case its finds the `lastModules` and then calls `prefect` async. 

So in this case, you could add tests to `Compiler.test.js` (@sokra correct me if I'm wrong).

**Comment by sokra** [comment]
As the `AutomaticPrefetchPlugin` depends on being in watch mode I would go with writing a `test/watchCases` test.

**Comment by alistairjcbrown** [comment]
The above feel suggestions feel like integration tests (ie. does the plugin work when called as part of a larger use case). As well as those, it might be nice to have unit tests too - there's an `applyPluginWithOptions` helper that's just been merged in #3684 (that PR has a test added too as an example of how to use it, or see #3685).
With that helper, you can assert a) does the plugin listen for the expected events and b) do the corresponding event handlers work as expected.

File: lib/AutomaticPrefetchPlugin.js [hunk 1]
```diff
@@ -2,34 +2,35 @@
...
-			return m instanceof NormalModule;
-		}).map(function(m) {
-			return {
-				context: m.context,
-				request: m.request
-			};
+			compilation.dependencyFactories.set(PrefetchDependency, normalModuleFactory);
+		});
+		let lastModules = null;
+		compiler.plugin("after-compile", (compilation, callback) => {
```
Comment (Kovensky) [inline lib/AutomaticPrefetchPlugin.js]: Would be nice if Tapable could be upgraded to support `Promise` by checking the `Function#length`, to see if the function even attempts to receive the callback argument :)

Idea from https://github.com/mochajs/mocha/blob/master/lib/runnable.js#L52 and https://github.com/mochajs/mocha/blob/master/lib/runnable.js#L302-L324, but it's easier for mocha to support that as it only ever gives the callback itself as an argument (i.e. promise-returning functions should have empty parameter lists, not necessarily the case with Tapable).

Comment (sokra) [inline lib/AutomaticPrefetchPlugin.js]: Supporting both has probably negative effect on performance, but I think it's possible to update everything to Promises (in future)...

**Comment by TheLarkInn** [comment]
Going to merge and then create an issue for adding test coverage for this plugin.

**Comment by TheLarkInn** [comment]
@chicoxyzzy Thank you!

**Pull request merged**